plugins/optimization-detective/class-od-link-collection.php

```diff
@@ -248,9 +248,33 @@ public function get_response_header(): ?string {
 		$link_headers = array();
 
 		foreach ( $this->get_prepared_links() as $link ) {
-			// The about:blank is present since a Link without a reference-uri is invalid so any imagesrcset would otherwise not get downloaded.
-			$link['href'] = isset( $link['href'] ) ? esc_url_raw( $link['href'] ) : 'about:blank';
-			$link_header  = '<' . $link['href'] . '>';
+			// Encode only the filename part of the URL to ensure it contains only ASCII characters.
+			if ( isset( $link['href'] ) ) {
+				$parsed_url = wp_parse_url( $link['href'] );
+				if ( isset( $parsed_url['path'] ) ) {
+					$path_segments        = explode( '/', $parsed_url['path'] );
+					$last_segment         = array_pop( $path_segments );
+					$encoded_last_segment = rawurlencode( $last_segment );
+
+					$encoded_path = implode( '/', $path_segments ) . '/' . $encoded_last_segment;
+
+					$scheme = isset( $parsed_url['scheme'] ) ? $parsed_url['scheme'] : '';
+					$host   = isset( $parsed_url['host'] ) ? $parsed_url['host'] : '';
+
+					$link['href'] = esc_url_raw( $scheme . '://' . $host . $encoded_path );
+
+					// Append query and fragment if they exist.
+					$link['href'] .= isset( $parsed_url['query'] ) ? '?' . $parsed_url['query'] : '';
+					$link['href'] .= isset( $parsed_url['fragment'] ) ? '#' . $parsed_url['fragment'] : '';
+				} else {
+					$link['href'] = esc_url_raw( $link['href'] );
+				}
```

Comment:
Instead of parsing the URL and then re-constructing it, what if you just check if the `href` has any non-ASCII chars and then encode the entire URL?

```suggestion
				if ( 1 === preg_match( '/[^\x20-\x7E]/', $link['href'] ) ) {
					$link['href'] = rawurlencode( urldecode( $link['href'] ) );
				} else {
					$link['href'] = esc_url_raw( $link['href'] );
				}
```

The regular expression range there is to match everything from a space to a tilde. This might not be the right range.

Reply:
Changed the regex to match any non-ascii characters